### PR TITLE
[Relay][Prelude] Use the Relay parser to define the Relay prelude

### PR DESCRIPTION
**Pull request by jroesch**
The current Relay prelude is huge pain to modify and doesn't always result in correct ASTs requiring painful hand modifications each time we update it.

It would be far more useful to make use of the Relay parser to load the Prelude from disk.

This PR makes it possible to partially load a Relay Prelude from disk. The PR also includes some modifications to the Relay parser to support defining more code in Relay.

This should help motivate one piece of the final push on the Relay Text format (see #3016).

My goal is to make it possible to define a limited form of TensorArray using this in the coming days.

cc @wweic @joshpoll @MarisaKirisame

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
AFAICT relay dont has a way of defining adt.
@joshpoll maybe we can add that?

**Comment by slyubomirsky** [comment]
Finalizing a Relay ADT syntax would be of great help to docs too

File: tests/python/relay/test_prelude.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import tvm
+from tvm import relay
+
+def load_prelude():
+    mod = relay.Module()
+    return relay.prelude.Prelude(mod)
+
+def test_list():
```
Comment (weberlo) [inline tests/python/relay/test_prelude.py]: Is this test supposed to do more?

File: python/tvm/relay/prelude.py [hunk 2]
```diff
@@ -534,6 +538,19 @@ def __init__(self, mod):
...
+
+__PRELUDE_PATH__ = os.path.dirname(os.path.realpath(__file__))
+
+def load_prelude(mod=None):
+    from .parser import fromtext
+
+    if mod is None:
+        mod = Module({})
+
+    file = os.path.join(__PRELUDE_PATH__, "prelude.rly")
```
Comment (weberlo) [inline python/tvm/relay/prelude.py]: `file` is a builtin. rename

File: python/tvm/relay/_parser.py [hunk 3]
```diff
@@ -454,15 +467,21 @@ def visitIncompleteType(self, ctx):
...
-                return ty.scalar_type(ident_type)
+            if type_ident.startswith(type_prefix):
+                return ty.scalar_type(type_ident)
+
+        type_param = lookup(self.type_param_scopes, type_ident)
+        if type_param:
+            return type_param
 
-        raise ParseError("Unknown builtin type: {}".format(ident_type))
+        print(type_ident)
```
Comment (weberlo) [inline python/tvm/relay/_parser.py]: remove

File: python/tvm/relay/_parser.py [hunk 4]
```diff
@@ -454,15 +467,21 @@ def visitIncompleteType(self, ctx):
         return None
 
     def visitIdentType(self, ctx):
+        # TODO(@jroesch) Identifier Type doesn't make sense should be Type Identifier.
```
Comment (weberlo) [inline python/tvm/relay/_parser.py]: Can we resolve this TODO in this PR?

File: python/tvm/relay/_parser.py [hunk 5]
```diff
@@ -368,14 +370,25 @@ def mk_func(self, ctx):
...
+            assert type_params
+            for ty_param in type_params:
+                name = ty_param.getText()
+                self.mk_typ(name, ty.Kind.Type)
+
         var_list, attr_list = self.visit(ctx.argList())
         ret_type = self.getType_(ctx.type_())
 
+        body = self.visit(ctx.body())
+        # NB(@jroesch): you must stay in the type parameter scope until
```
Comment (weberlo) [inline python/tvm/relay/_parser.py]: An "NB" shouldn't have a username, should it?

File: python/tvm/relay/grammar/Relay.g4 [hunk 6]
```diff
@@ -111,8 +111,8 @@ expr
   // | 'debug'                                # debug
   ;
 
-func: 'fn'        '(' argList ')' ('->' type_)? body ;
-defn: 'def' ident '(' argList ')' ('->' type_)? body ;
+func: 'fn'        (typeParamSeq)? '(' argList ')' ('->' type_)? body ;
```
Comment (joshpoll) [inline python/tvm/relay/grammar/Relay.g4]:
```suggestion
func: 'fn'        typeParamSeq? '(' argList ')' ('->' type_)? body ;
```

File: python/tvm/relay/grammar/Relay.g4 [hunk 7]
```diff
@@ -111,8 +111,8 @@ expr
...
+defn: 'def' ident (typeParamSeq)? '(' argList ')' ('->' type_)? body ;
```
Comment (joshpoll) [inline python/tvm/relay/grammar/Relay.g4]:
```suggestion
defn: 'def' ident typeParamSeq? '(' argList ')' ('->' type_)? body ;
```

File: python/tvm/relay/grammar/Relay.g4 [hunk 8]
```diff
@@ -140,7 +146,7 @@ type_
   | 'Tensor' '[' shapeSeq ',' type_ ']'             # tensorType
   // currently unused
   // | identType '[' (type_ (',' type_)*)? ']'         # callType
-  | 'fn' '(' (type_ (',' type_)*)? ')' '->' type_   # funcType
+  | 'fn' (typeParamSeq)? '(' (type_ (',' type_)*)? ')' '->' type_   # funcType
```
Comment (joshpoll) [inline python/tvm/relay/grammar/Relay.g4]:
```suggestion
  | 'fn' typeParamSeq? '(' (type_ (',' type_)*)? ')' '->' type_   # funcType
```

File: tests/python/relay/test_prelude.py (hunk 1 above)
Comment (wweic) [reply tests/python/relay/test_prelude.py]: We don't seem to need this test file as long as we migrate all built n functions and ADTs to using this mechanism. `test_adt.py` needs to use Prelude so the tests there will cover this change as wel..

File: python/tvm/relay/prelude.rly [hunk 9]
```diff
@@ -0,0 +1,11 @@
+v0.0.1
```
Comment (wweic) [inline python/tvm/relay/prelude.rly]: Try moving more functions here?

File: python/tvm/relay/_parser.py (hunk 5 above)
Comment (jroesch) [reply python/tvm/relay/_parser.py]: I like usernames so we can know how made the warning.

File: tests/python/relay/test_prelude.py (hunk 1 above)
Comment (jroesch) [reply tests/python/relay/test_prelude.py]: I think we should incrementally move over. It is too much work to do in one pr.

**Comment by slyubomirsky** [comment]
@jroesch has asked me to take over this PR so I will work on addressing feedback and fixing linting errors to get this to a mergeable state soon. The next step would be to add ADTs and matching to the text format so the entire prelude can be written in the text format.

File: python/tvm/relay/_parser.py (hunk 4 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Will do

File: python/tvm/relay/grammar/Relay.g4 (hunk 8 above)
Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Done

File: python/tvm/relay/grammar/Relay.g4 (hunk 7 above)
Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Done

File: python/tvm/relay/grammar/Relay.g4 (hunk 6 above)
Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Done

File: python/tvm/relay/_parser.py (hunk 4 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Done

File: python/tvm/relay/_parser.py (hunk 3 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Done

File: python/tvm/relay/prelude.py (hunk 2 above)
Comment (slyubomirsky) [reply python/tvm/relay/prelude.py]: Done

**Comment by slyubomirsky** [comment]
Pinging reviewers: @wweic @joshpoll @MarisaKirisame

File: python/tvm/relay/prelude.rly (hunk 9 above)
Comment (slyubomirsky) [reply python/tvm/relay/prelude.rly]: We'd need to add ADTs and matching to the text format, which I think would be better left for another PR. It's good to have a simple core merged ahead of the rest, as that will make it easier to iterate afterwards (others can contribute too)

File: python/tvm/relay/grammar/Relay.g4 [hunk 10]
```diff
@@ -19,7 +19,7 @@
 
 grammar Relay;
 
-SEMVER: 'v0.0.1' ;
+SEMVER: 'v0.0.1.1' ;
```
Comment (joshpoll) [inline python/tvm/relay/grammar/Relay.g4]:
```suggestion
SEMVER: 'v0.0.2' ;
```

This is the convention for pre-release semver.

File: python/tvm/relay/grammar/Relay.g4 [hunk 11]
```diff
@@ -132,15 +132,21 @@ attr: CNAME '=' expr ;
 // relations: 'where' relation (',' relation)* ;
 // relation: ident '(' (type_ (',' type_)*)? ')' ;
 
+typeParamSeq
+  : '[' ']'
+  | '[' ident ']'
+  | '[' ident (',' ident)+ ']'
```
Comment (joshpoll) [inline python/tvm/relay/grammar/Relay.g4]:
```suggestion
  | '[' ident (',' ident)* ']'
```
to merge this case and the previous one.

Comment (joshpoll) [reply python/tvm/relay/grammar/Relay.g4]: tuples need something different b/c they require a trailing comma

File: tests/python/relay/test_ir_parser.py [hunk 12]
```diff
@@ -28,7 +28,7 @@
 else:
     raises_parse_error = lambda x: x
 
-SEMVER = "v0.0.1"
+SEMVER = "v0.0.1.1"
```
Comment (joshpoll) [inline tests/python/relay/test_ir_parser.py]:
```suggestion
SEMVER = "v0.0.2"
```

File: python/tvm/relay/_parser.py [hunk 13]
```diff
@@ -453,16 +466,23 @@ def visitIncompleteType(self, ctx):
         # type (RelayParser.IncompleteTypeContext) -> None:
         return None
 
-    def visitIdentType(self, ctx):
-        # type: (RelayParser.IdentTypeContext) -> Union[ty.TensorType, str]
-        ident_type = ctx.CNAME().getText()
+    def visitTypeIdent(self, ctx):
+        '''
+        Handle type identifier.
+        type: (RelayParser.TypeIdentContext) -> Union[ty.TensorType, str]
```
Comment (joshpoll) [inline python/tvm/relay/_parser.py]: this type annotation should be `# type: (RelayParser.IdentTypeContext) -> Union[ty.TensorType, str]` and go above and outside the doc string. cf: https://mypy.readthedocs.io/en/latest/cheat_sheet.html

File: python/tvm/relay/_parser.py [hunk 14]
```diff
@@ -453,16 +466,23 @@ def visitIncompleteType(self, ctx):
...
-        # look through all type prefixes for a match
+        # Look through all type prefixes for a match
         for type_prefix in TYPE_PREFIXES:
-            if ident_type.startswith(type_prefix):
-                return ty.scalar_type(ident_type)
+            if type_ident.startswith(type_prefix):
+                return ty.scalar_type(type_ident)
+
+        type_param = lookup(self.type_param_scopes, type_ident)
+        if type_param:
```
Comment (joshpoll) [inline python/tvm/relay/_parser.py]:
```suggestion
        if type_param is not None:
```

File: python/tvm/relay/prelude.py [hunk 15]
```diff
@@ -534,6 +538,22 @@ def __init__(self, mod):
...
         self.define_tree_size()
 
-        self.define_id()
-        self.define_compose()
         self.define_iterate()
+
+__PRELUDE_PATH__ = os.path.dirname(os.path.realpath(__file__))
+
+def load_prelude(mod=None):
+    '''Parses the portions of the Prelude written in Relay's text format and adds
```
Comment (joshpoll) [inline python/tvm/relay/prelude.py]: newline after `'''`. Consider using either `'''` or `"""` for all docstrings in this file

File: python/tvm/relay/_parser.py (hunk 13 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Pylint complained when I had it that way

Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Ah, it complained because I had the annotation after the docstring, not before

**Comment by slyubomirsky** [comment]
Thanks for the feedback. I also removed `test_prelude` because it is redundant, since `test_ast` tests the prelude functions, which would require loading the file

File: python/tvm/relay/grammar/Relay.g4 (hunk 10 above)
Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Thanks, done

File: python/tvm/relay/grammar/Relay.g4 (hunk 11 above)
Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Good idea. Done

File: python/tvm/relay/_parser.py (hunk 13 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Done

File: python/tvm/relay/_parser.py (hunk 14 above)
Comment (slyubomirsky) [reply python/tvm/relay/_parser.py]: Thanks, done

File: python/tvm/relay/prelude.py (hunk 15 above)
Comment (slyubomirsky) [reply python/tvm/relay/prelude.py]: Done

**Comment by slyubomirsky** [comment]
Fixing the previous test failure led me to realize that we have an issue with the parser flag: If we want to have the Prelude and other important Relay functionality written using the text format, should the parser be an optional flag? I changed the prelude to check whether the parser is enabled and default to the old manually written ASTs, but this does not seem like a particularly good practice, since it requires us to maintain two definitions, one of which is indeed a manually assembled AST. What do you all believe is a good approach moving forward? (addressed to the reviewers, also @jroesch and @tqchen)

**Comment by jroesch** [comment]
We should conditionally support generating the parser, when we check in we should generate a copy, commit it, and mark the files as binaries in my opinion, it satisfies both design tensions.

**Comment by slyubomirsky** [comment]
I agree that there should be a pre-generated parser available for users who do not intend to modify the grammar or parser. Does everyone agree that we should have a pre-generated parser committed and marked as a binary?

**Comment by slyubomirsky** [comment]
I've now added the parser to the repo so users can always run the parser, even if they don't have ANTLR enabled in their build. It will be easy to revert that commit if this choice is objectionable

**Comment by wweic** [comment]
I think having a pre-generated parser is a good idea, it will also enable running text format unit test for every tvm user without ANTLR installed.

**Comment by slyubomirsky** [comment]
Would appreciate input from anyone knowledgeable whether the generated parser files _should_ have the Apache headers on them (we'd need to script for them to be appended during the build if so)

File: python/tvm/relay/grammar/py2/RelayLexer.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
+# Generated from /home/sslyu/tvm/python/tvm/relay/grammar/Relay.g4 by ANTLR 4.7.2
```
Comment (jroesch) [inline python/tvm/relay/grammar/py2/RelayLexer.py]: Can you mark these all as binary files?

Comment (slyubomirsky) [reply python/tvm/relay/grammar/py2/RelayLexer.py]: I did in .gitattributes, but it didn't change how they display on Github. I will try this as well: https://help.github.com/en/articles/customizing-how-changed-files-appear-on-github

File: tests/python/relay/test_ir_parser.py [hunk 17]
```diff
@@ -113,7 +98,7 @@ def test_comments():
         UNIT
     )
 
-@if_parser_enabled
+
 def test_int_literal():
     assert isinstance(relay.fromtext(SEMVER+"1"), relay.Constant)
     assert isinstance(relay.fromtext(SEMVER+"1").data, tvm.ndarray.NDArray)
```
Comment (weberlo) [inline tests/python/relay/test_ir_parser.py]: I know this isn't the point of this PR, but can we have `relay.fromtext(SEMVER + _)` as a helper function for this test file?  It's a bit of an eye sore.

Comment (slyubomirsky) [reply tests/python/relay/test_ir_parser.py]: Done

**Comment by slyubomirsky** [comment]
No idea what's with the test failure but it definitely appears unrelated

**Pull request merged**